src/jdk.compiler/share/classes/com/sun/tools/javac/tree/TreeInfo.java

```diff
@@ -1352,42 +1344,40 @@ public static boolean isErrorEnumSwitch(JCExpression selector, List<JCCase> case
                              .allMatch(p -> p.hasTag(IDENT));
     }
 
-    public static PatternPrimaryType primaryPatternType(JCPattern pat) {
+    public static PatternPrimaryType primaryPatternType(JCTree pat) {
         return switch (pat.getTag()) {
-            case BINDINGPATTERN -> new PatternPrimaryType(((JCBindingPattern) pat).type, true);
-            case GUARDPATTERN -> {
-                JCGuardPattern guarded = (JCGuardPattern) pat;
-                PatternPrimaryType nested = primaryPatternType(guarded.patt);
-                boolean unconditional = nested.unconditional();
-                if (guarded.expr.type.hasTag(BOOLEAN) && unconditional) {
-                    unconditional = false;
-                    var constValue = guarded.expr.type.constValue();
-                    if (constValue != null && ((int) constValue) == 1) {
-                        unconditional = true;
-                    }
-                }
-                yield new PatternPrimaryType(nested.type(), unconditional);
-            }
+            case BINDINGPATTERN -> new PatternPrimaryType(((JCBindingPattern) pat).type);
             case PARENTHESIZEDPATTERN -> primaryPatternType(((JCParenthesizedPattern) pat).pattern);
             default -> throw new AssertionError();
         };
     }
 
-    public static JCBindingPattern primaryPatternTree(JCPattern pat) {
+    public static JCBindingPattern primaryPatternTree(JCTree pat) {
         return switch (pat.getTag()) {
             case BINDINGPATTERN -> (JCBindingPattern) pat;
-            case GUARDPATTERN -> primaryPatternTree(((JCGuardPattern) pat).patt);
             case PARENTHESIZEDPATTERN -> primaryPatternTree(((JCParenthesizedPattern) pat).pattern);
             default -> throw new AssertionError();
         };
     }
 
-    public record PatternPrimaryType(Type type, boolean unconditional) {}
+    public record PatternPrimaryType(Type type) {}
 
     public static boolean expectedExhaustive(JCSwitch tree) {
         return tree.patternSwitch ||
                tree.cases.stream()
                          .flatMap(c -> c.labels.stream())
                          .anyMatch(l -> TreeInfo.isNull(l));
     }
+
+    public static boolean unconditionalCaseLabel(JCCaseLabel cse) {
+        if (!cse.isPattern()) {
+            return true;
+        }
+        JCExpression guard = ((JCPattern) cse).guard;
+        if (guard != null && guard.type.hasTag(BOOLEAN)) {
```

Comment:
In the spec we see the following:

> A pattern case element is said to be unrefined if either (i) it has no associated when expression, or (ii) it has an associated when expression that is a constant expression ([15.29](https://docs.oracle.com/javase/specs/jls/se18/html/jls-15.html#jls-15.29)) with value true. A case element is unrefined if it is not a pattern case element or it is an unrefined pattern case element.

However, in the definition of `unconditional` I don't see the part for the constant expression.

I wonder if the spec should be updated or the code at this point

Reply:
the code below checks that the guard is boolean constant `true`, this is the way the compiler can do that, a constant can be retrieved using the `constValue` method